core/db/migrate/20141217215630_update_product_slug_index.rb

```diff
@@ -0,0 +1,6 @@
+class UpdateProductSlugIndex < ActiveRecord::Migration
+  def change
+    remove_index :spree_products, :slug
+    add_index :spree_products, :slug, unique: true
```

Comment:
Unfortunately can't use partial indexes because of MySQL :cry:. This is worked around by using a `punch_slug` method on Product. It's not ideal, but I'm not looking to change that in this PR, just merge the two indexes.